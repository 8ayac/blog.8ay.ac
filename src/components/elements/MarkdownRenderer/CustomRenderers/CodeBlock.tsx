import React from 'react';
import { Prism } from 'react-syntax-highlighter';
import { atomDark } from 'react-syntax-highlighter/dist/cjs/styles/prism';

export const CodeBlock: React.FC<{ language: string; value: string }> = ({
  language,
  value,
}) => (
  <Prism
    style={atomDark}
    language={language}
    customStyle={{
      padding: '1.5rem 2rem',
      borderRadius: '5px',
      fontFamily: 'monospace',
      lineHeight: '1.1',
    }}
  >
    {value}
  </Prism>
);
