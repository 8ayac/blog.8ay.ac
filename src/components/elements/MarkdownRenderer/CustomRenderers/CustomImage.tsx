import { config } from '@/site.config';
import Image from 'next/image';
import React from 'react';
import urljoin from 'url-join';

export const CustomImage: React.FC<{
  src: string;
  alt?: string;
  title?: string;
}> = ({ src, alt, title }) => {
  const srcUrl = /^https?:\/\/.*/.test(src)
    ? src
    : urljoin(config.site.rootUrl, src);
  const { searchParams } = new URL(srcUrl);
  const width = searchParams.get('w');
  const height = searchParams.get('h');

  return (
    <>
      {width && height ? (
        <Image
          src={src}
          alt={alt}
          title={title}
          width={`${width}px`}
          height={`${height}px`}
        />
      ) : (
        <img src={src} alt={alt} title={title} />
      )}
    </>
  );
};
