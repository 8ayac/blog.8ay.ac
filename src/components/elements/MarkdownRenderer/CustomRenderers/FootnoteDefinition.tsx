import {
  FOOTNOTE_DEFINITION_PREFIX,
  FOOTNOTE_REFERENCE_PREFIX,
} from '@/src/constants/footnote';
import styled from '@emotion/styled';
import React, { ReactElement } from 'react';

const S = {
  FootnoteWrapperDiv: styled.div`
    margin-bottom: 1rem !important;

    * {
      display: inline;
    }
  `,

  LabelSpan: styled.span`
    margin-right: 1rem;
    font-weight: 600;
  `,

  ReturnAnchor: styled.a`
    margin: 0 0.2rem !important;
  `,
};

export const FootnoteDefinition: React.FC<{
  identifier: string;
  label: string;
  children: ReactElement;
}> = ({ identifier, label, children }) => {
  return (
    <>
      <S.FootnoteWrapperDiv id={`${FOOTNOTE_DEFINITION_PREFIX}${identifier}`}>
        <S.LabelSpan>^{label}:</S.LabelSpan>
        {children}
        <S.ReturnAnchor href={`#${FOOTNOTE_REFERENCE_PREFIX}${identifier}`}>
          â†µ
        </S.ReturnAnchor>
      </S.FootnoteWrapperDiv>
    </>
  );
};
