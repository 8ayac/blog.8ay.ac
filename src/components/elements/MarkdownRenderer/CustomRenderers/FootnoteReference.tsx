import {
  FOOTNOTE_DEFINITION_PREFIX,
  FOOTNOTE_REFERENCE_PREFIX,
} from '@/src/constants/footnote';
import styled from '@emotion/styled';
import React from 'react';

const S = {
  IdentifierSup: styled.sup`
    margin: 0.2rem;
  `,
};

export const FootnoteReference: React.FC<{
  identifier: string;
  label: string;
}> = ({ identifier, label }) => {
  return (
    <S.IdentifierSup id={`${FOOTNOTE_REFERENCE_PREFIX}${identifier}`}>
      <a href={`#${FOOTNOTE_DEFINITION_PREFIX}${identifier}`}>{label}</a>
    </S.IdentifierSup>
  );
};
