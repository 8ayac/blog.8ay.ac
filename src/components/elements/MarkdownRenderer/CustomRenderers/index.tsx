import { CustomImage } from '@/src/components/elements/MarkdownRenderer/CustomRenderers/CustomImage';
import { FootnoteDefinition } from '@/src/components/elements/MarkdownRenderer/CustomRenderers/FootnoteDefinition';
import { FootnoteReference } from '@/src/components/elements/MarkdownRenderer/CustomRenderers/FootnoteReference';

export const customRenderers = {
  footnoteDefinition: FootnoteDefinition,
  footnoteReference: FootnoteReference,
  image: CustomImage,
};
