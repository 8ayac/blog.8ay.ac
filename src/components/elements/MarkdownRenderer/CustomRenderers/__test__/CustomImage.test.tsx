import { CustomImage } from '@/src/components/elements/MarkdownRenderer/CustomRenderers/CustomImage';
import { mount } from 'enzyme';
import React from 'react';

describe('CustomImage', () => {
  const TEST_WIDTH = 100;
  const TEST_HEIGHT = 200;

  const testPropsSrcIsRelativePath = {
    src: `/a.png?w=${TEST_WIDTH}&h=${TEST_HEIGHT}`,
    alt: 'this is a test image',
    title: 'Test Image',
  };
  const testPropsSrcIsURL = {
    src: `http://example.com/a.png?w=${TEST_WIDTH}&h=${TEST_HEIGHT}`,
    alt: 'this is a test image',
    title: 'Test Image',
  };
  const testPropsWithoutImageWidthAndHeight = {
    src: '/a.png',
    alt: 'this is a test image',
    title: 'Test Image',
  };

  describe('renders images correctly', () => {
    describe('to match the snapshot', () => {
      test("when setting a relative image path to the value of 'src'", () => {
        const wrapper = mount(<CustomImage {...testPropsSrcIsRelativePath} />);
        expect(wrapper).toMatchSnapshot();
      });

      test("when setting image URL to the value of 'src'", () => {
        const wrapper = mount(<CustomImage {...testPropsSrcIsURL} />);
        expect(wrapper).toMatchSnapshot();
      });

      test("when setting the image path without width and height to the value of 'src'", () => {
        const wrapper = mount(
          <CustomImage {...testPropsWithoutImageWidthAndHeight} />
        );
        expect(wrapper).toMatchSnapshot();
      });
    });

    describe('with specified sizes', () => {
      test("when setting a relative image path to the value of 'src'", () => {
        const wrapper = mount(<CustomImage {...testPropsSrcIsRelativePath} />);
        expect(wrapper.find('Image').prop('width')).toBe(`${TEST_WIDTH}px`);
        expect(wrapper.find('Image').prop('height')).toBe(`${TEST_HEIGHT}px`);
      });

      test("when setting image URL to the value of 'src'", () => {
        const wrapper = mount(<CustomImage {...testPropsSrcIsURL} />);
        expect(wrapper.find('Image').prop('width')).toBe(`${TEST_WIDTH}px`);
        expect(wrapper.find('Image').prop('height')).toBe(`${TEST_HEIGHT}px`);
      });

      test("when setting the image path without width and height to the value of 'src'", () => {
        const wrapper = mount(
          <CustomImage {...testPropsWithoutImageWidthAndHeight} />
        );
        expect(wrapper.find('Image')).toHaveLength(0);
      });
    });

    describe('with proper attributes', () => {
      test("when setting a relative image path to the value of 'src'", () => {
        const wrapper = mount(<CustomImage {...testPropsSrcIsRelativePath} />);
        expect(wrapper.find('Image').prop('src')).toBe(
          testPropsSrcIsRelativePath.src
        );
        expect(wrapper.find('Image').prop('alt')).toBe(
          testPropsSrcIsRelativePath.alt
        );
        expect(wrapper.find('Image').prop('title')).toBe(
          testPropsSrcIsRelativePath.title
        );

        expect(wrapper.find('Image').prop('width')).toBe(`${TEST_WIDTH}px`);
        expect(wrapper.find('Image').prop('height')).toBe(`${TEST_HEIGHT}px`);
      });
    });

    test("when setting image URL to the value of 'src'", () => {
      const wrapper = mount(<CustomImage {...testPropsSrcIsURL} />);
      expect(wrapper.find('Image').prop('src')).toBe(testPropsSrcIsURL.src);
      expect(wrapper.find('Image').prop('alt')).toBe(testPropsSrcIsURL.alt);
      expect(wrapper.find('Image').prop('title')).toBe(testPropsSrcIsURL.title);

      expect(wrapper.find('Image').prop('width')).toBe(`${TEST_WIDTH}px`);
      expect(wrapper.find('Image').prop('height')).toBe(`${TEST_HEIGHT}px`);
    });

    test("when setting the image path without width and height to the value of 'src'", () => {
      const wrapper = mount(
        <CustomImage {...testPropsWithoutImageWidthAndHeight} />
      );
      expect(wrapper.find('Image')).toHaveLength(0);

      expect(wrapper.find('img').prop('src')).toBe(
        testPropsWithoutImageWidthAndHeight.src
      );
      expect(wrapper.find('img').prop('alt')).toBe(
        testPropsWithoutImageWidthAndHeight.alt
      );
      expect(wrapper.find('img').prop('title')).toBe(
        testPropsWithoutImageWidthAndHeight.title
      );
    });
  });
});
