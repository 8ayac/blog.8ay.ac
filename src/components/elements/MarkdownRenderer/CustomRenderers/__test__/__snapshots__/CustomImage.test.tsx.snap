// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomImage renders images correctly to match the snapshot when setting a relative image path to the value of 'src' 1`] = `
<CustomImage
  alt="this is a test image"
  src="/a.png?w=100&h=200"
  title="Test Image"
>
  <Image
    alt="this is a test image"
    height="200px"
    src="/a.png?w=100&h=200"
    title="Test Image"
    width="100px"
  >
    <div
      style={
        Object {
          "boxSizing": "border-box",
          "display": "inline-block",
          "margin": 0,
          "maxWidth": "100%",
          "overflow": "hidden",
          "position": "relative",
        }
      }
    >
      <div
        style={
          Object {
            "boxSizing": "border-box",
            "display": "block",
            "maxWidth": "100%",
          }
        }
      >
        <img
          alt=""
          aria-hidden={true}
          role="presentation"
          src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTAwIiBoZWlnaHQ9IjIwMCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
          style={
            Object {
              "border": "none",
              "display": "block",
              "margin": 0,
              "maxWidth": "100%",
              "padding": 0,
            }
          }
        />
      </div>
      <img
        alt="this is a test image"
        decoding="async"
        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
        style={
          Object {
            "border": "none",
            "bottom": 0,
            "boxSizing": "border-box",
            "display": "block",
            "height": 0,
            "left": 0,
            "margin": "auto",
            "maxHeight": "100%",
            "maxWidth": "100%",
            "minHeight": "100%",
            "minWidth": "100%",
            "objectFit": undefined,
            "objectPosition": undefined,
            "padding": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "visibility": "hidden",
            "width": 0,
          }
        }
        title="Test Image"
      />
    </div>
  </Image>
</CustomImage>
`;

exports[`CustomImage renders images correctly to match the snapshot when setting image URL to the value of 'src' 1`] = `
<CustomImage
  alt="this is a test image"
  src="http://example.com/a.png?w=100&h=200"
  title="Test Image"
>
  <Image
    alt="this is a test image"
    height="200px"
    src="http://example.com/a.png?w=100&h=200"
    title="Test Image"
    width="100px"
  >
    <div
      style={
        Object {
          "boxSizing": "border-box",
          "display": "inline-block",
          "margin": 0,
          "maxWidth": "100%",
          "overflow": "hidden",
          "position": "relative",
        }
      }
    >
      <div
        style={
          Object {
            "boxSizing": "border-box",
            "display": "block",
            "maxWidth": "100%",
          }
        }
      >
        <img
          alt=""
          aria-hidden={true}
          role="presentation"
          src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTAwIiBoZWlnaHQ9IjIwMCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
          style={
            Object {
              "border": "none",
              "display": "block",
              "margin": 0,
              "maxWidth": "100%",
              "padding": 0,
            }
          }
        />
      </div>
      <img
        alt="this is a test image"
        decoding="async"
        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
        style={
          Object {
            "border": "none",
            "bottom": 0,
            "boxSizing": "border-box",
            "display": "block",
            "height": 0,
            "left": 0,
            "margin": "auto",
            "maxHeight": "100%",
            "maxWidth": "100%",
            "minHeight": "100%",
            "minWidth": "100%",
            "objectFit": undefined,
            "objectPosition": undefined,
            "padding": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "visibility": "hidden",
            "width": 0,
          }
        }
        title="Test Image"
      />
    </div>
  </Image>
</CustomImage>
`;

exports[`CustomImage renders images correctly to match the snapshot when setting the image path without width and height to the value of 'src' 1`] = `
<CustomImage
  alt="this is a test image"
  src="/a.png"
  title="Test Image"
>
  <img
    alt="this is a test image"
    src="/a.png"
    title="Test Image"
  />
</CustomImage>
`;
